src/tools/wasm-link.cc

```diff
@@ -504,14 +504,22 @@ static void write_names_section(Context* ctx) {
   WRITE_UNKNOWN_SIZE(stream);
   write_u32_leb128(stream, total_count, "element count");
   for (const std::unique_ptr<LinkerInputBinary>& binary: ctx->inputs) {
+    size_t delta = 0;
+    uint32_t off;
     for (size_t i = 0; i < binary->debug_names.size(); i++) {
       if (binary->debug_names[i].empty())
         continue;
       if (i < binary->function_imports.size()) {
-        if (!binary->function_imports[i].active)
+        if (!binary->function_imports[i].active) {
+          delta++;
           continue;
+        }
+        off = binary->imported_function_index_offset - delta;
+      } else {
+        off = binary->function_index_offset;
       }
-      write_u32_leb128(stream, i + binary->function_index_offset, "function index");
+
```

Comment:
I don't think that deals with inactive imports? I must confess I don't fully understand the code, so it's probably best to use `relocate_func_index` and fix that (if necessary).